crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,781 @@
+                (
+                    hashable_params(&first_instr.params),
+                    hashable_params(&second_instr.params),
+                ),
+            )) {
+                self._cache_hit += 1;
+                return *commutation;
+            } else {
+                self._cache_miss += 1;
+            }
+        } else {
+            self._cache_miss += 1;
+        }
+
+        // perform matrix multiplication to determine commutation
+        let is_commuting =
+            self.commute_matmul(first_instr, first_qargs, second_instr, second_qargs);
+
+        // TODO: implement a LRU cache for this
+        if self.current_cache_entries >= self.cache_max_entries {
+            self.clear_cache();
+        }
+        // cache results from is_commuting
+        self.cache
+            .entry((first_op.name().to_string(), second_op.name().to_string()))
+            .and_modify(|entries| {
+                let key = (
+                    Self::get_relative_placement(first_qargs, second_qargs),
+                    (
+                        hashable_params(&first_instr.params),
+                        hashable_params(&second_instr.params),
+                    ),
+                );
+                entries.insert(key, is_commuting);
+                self.current_cache_entries += 1;
+            })
+            .or_insert_with(|| {
+                let mut entries = HashMap::with_capacity(1);
+                let key = (
+                    Self::get_relative_placement(first_qargs, second_qargs),
+                    (
+                        hashable_params(&first_instr.params),
+                        hashable_params(&second_instr.params),
+                    ),
+                );
+                entries.insert(key, is_commuting);
+                self.current_cache_entries += 1;
+                CommutationCacheEntry { mapping: entries }
+            });
+        is_commuting
+    }
+
+    fn commute_matmul(
+        &self,
+        first_instr: &CircuitInstruction,
+        first_qargs: &[usize],
+        second_instr: &CircuitInstruction,
+        second_qargs: &[usize],
+    ) -> bool {
+        // compute relative positioning of qargs of the second gate to the first gate
+        let mut qarg: HashMap<&usize, usize> =
+            HashMap::with_capacity(first_qargs.len() + second_qargs.len());
+        for (i, q) in first_qargs.iter().enumerate() {
+            qarg.entry(q).or_insert(i);
+        }
+        let mut num_qubits = first_qargs.len();
+        for q in second_qargs {
+            if !qarg.contains_key(q) {
+                qarg.insert(q, num_qubits);
+                num_qubits += 1;
+            }
+        }
+
+        let first_qarg: Vec<_> = first_qargs
+            .iter()
+            .map(|q| *qarg.get_item(q).unwrap())
+            .collect();
+        let second_qarg: Vec<_> = second_qargs
+            .iter()
+            .map(|q| *qarg.get_item(q).unwrap())
+            .collect();
```

Comment:
I'm not sure if `get_item` is what you need here. `get_item` comes from the `rustworkx_core` trait `DistanceMap` and it allows you to measure the distance to a `Key` it might work well with tests but I'd suggest you use the `.get()` method.

```suggestion
        let first_qarg: Vec<_> = first_qargs.iter().map(|q| *qarg.get(q).unwrap()).collect();
        let second_qarg: Vec<_> = second_qargs.iter().map(|q| *qarg.get(q).unwrap()).collect();
```